### PR TITLE
Try to read data from closed watch before closing it

### PR DESCRIPTION
**Pull request by wojtek-t**
This is a similar PR to #31016

I have a feeling that if server is terminating watch, we are not reusing the connection.

@lavalamp

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32287)

### CODE REVIEW

File: pkg/util/framer/framer.go [hunk 1]
```diff
@@ -21,6 +21,7 @@ import (
 	"encoding/binary"
 	"encoding/json"
 	"io"
+	"io/ioutil"
```
Comment (wojtek-t) [inline pkg/util/framer/framer.go]: @smarterclayton @krousey - I'm not entirely sure this PR is even correct. But just throwing it to start discussing about it.

**Comment by wojtek-t** [comment]
@gmarek

**Comment by wojtek-t** [comment]
If this PR is correct, we may consider cherrypicking it to 1.4.

**Comment by krousey** [comment]
Two questions:
- Won't the server potentially just keep writing events?
- Can we even reuse a hijacked HTTP1 connection?

**Comment by krousey** [comment]
Or are watches not hijacked? I thought they were, but I can't find evidence of that now.

**Comment by smarterclayton** [comment]
Watches are not hijacked, just flushed.  

Will look - I thought we were pretty good at reading all the data here but I can imagine edge cases.

File: pkg/util/framer/framer.go [hunk 2]
```diff
@@ -105,6 +106,11 @@ func (r *lengthDelimitedFrameReader) Read(data []byte) (int, error) {
 }
 
 func (r *lengthDelimitedFrameReader) Close() error {
+	// The underlying ReadCloser is usually http response body.
+	// We try to ensure that it is fully read and closed before we
+	// reconnect, so that we reuse the same TCP connection.
+	const maxBodySlurpSize = 2 << 10
+	io.Copy(ioutil.Discard, &io.LimitedReader{R: r.r, N: maxBodySlurpSize})
```
Comment (smarterclayton) [inline pkg/util/framer/framer.go]: I feel like this should become a util function now - DiscardUpToLimit

**Comment by wojtek-t** [comment]
Hmm - this PR may not be correct - a bunch of tests failed here. Let me rerun them

@k8s-bot test this, issue: #IGNORE

**Comment by wojtek-t** [comment]
OK - so it's pretty clear that this PR is incorrect. However, I don't know why...

**Comment by smarterclayton** [comment]
I can't see anything obvious but somewhat brain fried right now.

**Comment by wojtek-t** [comment]
Let's leave it for now and get back to it after 1.4 is out.

**Comment by smarterclayton** [comment]
Can we close this?

**Comment by wojtek-t** [comment]
Yeah - I think we can.

**Pull request closed without merging**